### PR TITLE
Add `content::{Accept, ContentType}` headers

### PR DESCRIPTION
**Pull request by yoshuawuyts**
Adds the `content::{Accept, ContentType}` headers, implementing content negotation for media types. Supersedes https://github.com/http-rs/http-types/pull/269 which only implemented `content::ContentType`. Ref #99. Thanks!

### CODE REVIEW

**Review by Fishrock123 — CHANGES_REQUESTED** [review]
I don't think we should merge this until the API for content-type is flushed out more.

As far as I am aware, there are 4 main uses for `Content-Type`:
- Accept, i.e. negotiation. Covered in this PR.
- Determining file type from http data
- Determining http data from file and/or path
- Determining if a response should be compressed on-demand

The latter 3 are not covered here and are, in my experience, the more  necessary usage.

Here's some 23:00 ideas:
- `ext() -> &str` file extension as string
- `From<Path>` & `from_ext(&str)`
- `compressible() -> bool`

All of these require some kind of in-memory database. I was trying to figure out what structure to use for that but really wasn't sure. However we should populate that database from [mime-db](https://github.com/jshttp/mime-db), since it is by far the most comprehensive source out there (it is an aggregate plus more).

We could also do mime sniffing... but that's kinda costly, requires intercepting bodies, is possibly less reliable, etc

**Comment by yoshuawuyts** [comment]
@Fishrock123 To me it sounds like the features you're suggesting may be out of scope for this PR, even if they seem useful. This PR introduces two typed headers only:

- `Accept` which is used for content negotiation.
- `Content-Type` which is used to communicate the media type used.

Both headers implement the full spec, and in addition to that we also include the content negotiation algorithm to actually make `Accept` _useful_.

Your suggestions around `mime-db` likely best be implemented as extensions to our existing `http_types::mime::Mime` implementation. `Mime` already has some capabilities to infer the type either by reading magic bytes ([`Mime::sniff`](https://docs.rs/http-types/2.7.0/http_types/mime/struct.Mime.html#method.sniff)) or by attempting to guess the mime from the file extension ([`Mime::from_extension`](https://docs.rs/http-types/2.7.0/http_types/mime/struct.Mime.html#method.from_extension)).

We also already have a direct "create a mime from a file" conversion available through ([`Body::from_file`](https://docs.rs/http-types/2.7.0/http_types/struct.Body.html#method.from_file)), which applies both the `from_extension` and `sniff` methods to determine the file type. Finally `Mime` also has a `Display` impl that can be used to convert it back out to a string. Though a `Mime` -> file extension conversion may indeed be useful.

In short: I think your suggestions are good, but they don't need to block this PR. I've filed separate issues to track what's not been implemented so we can remember to implement them later:
- Add `Mime::into_extension`
- Add `Mime::is_compressible`
- Port `mime-db` to Rust to enhance our mime extension guessing
- Meta issue to track media type improvements

**Comment by Fishrock123** [comment]
> I don't think we should merge this until the API for content-type is flushed out more.

Ah, well, this and the issues from #274 are what I meant by this - i.e. that we have at least made an effort to think about it.

File: src/mime/mod.rs [hunk 1]
```diff
@@ -109,6 +109,33 @@ impl Mime {
             })
             .flatten()
     }
+
+    /// Remove a param from the set. Returns the `ParamValue` if it was contained within the set.
+    pub fn remove_param(&mut self, name: impl Into<ParamName>) -> Option<ParamValue> {
```
Comment (Fishrock123) [inline src/mime/mod.rs:114]: Should this be `take_param`? (Idk.)

File: src/content/content_type.rs [hunk 2]
```diff
@@ -0,0 +1,140 @@
+use std::{convert::TryInto, str::FromStr};
+
+use crate::headers::{HeaderName, HeaderValue, Headers, CONTENT_TYPE};
+use crate::Mime;
+
+/// Indicate the media type of the resource.
```
Comment (Fishrock123) [inline src/content/content_type.rs:6]:
```suggestion
/// Indicate the media type of a resource's content.
```

File: src/content/content_type.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+/// # Examples
+///
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::content::ContentType;
+/// use http_types::{Response, Mime};
+/// use std::str::FromStr;
+///
+/// let content_type = ContentType::new(Mime::from_str("text/*")?);
```
Comment (Fishrock123) [inline src/content/content_type.rs:24]: Can this be?:
```suggestion
/// let content_type = ContentType::new("text/*");
```

Or is the issue that `new()` here unwraps?

File: src/content/accept.rs [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+/// ```
+/// # fn main() -> http_types::Result<()> {
+/// #
+/// use http_types::content::{Accept, MediaTypeProposal};
+/// use http_types::{mime, Response};
+///
+/// let mut accept = Accept::new();
+/// accept.push(MediaTypeProposal::new(mime::HTML, Some(0.8))?);
+/// accept.push(MediaTypeProposal::new(mime::XML, Some(0.4))?);
+/// accept.push(MediaTypeProposal::new(mime::PLAIN, None)?);
```
Comment (Fishrock123) [inline src/content/accept.rs:37]:
```suggestion
/// accept.push(mime::PLAIN);
```

**Review by Fishrock123 — APPROVED** [review]
Overall looks good, with some minor fixes & questions.

File: src/mime/mod.rs (hunk 1 above)
Comment (yoshuawuyts) [reply src/mime/mod.rs:114]: This is modeled after [`HashMap::remove`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.remove) -- it removes the value; however if the entry existed it returns it.
